spring-integration-core/src/test/java/org/springframework/integration/handler/AbstractReplyProducingMessageHandlerTests.java

```diff
@@ -109,6 +108,33 @@ protected Object handleRequestMessage(Message<?> requestMessage) {
 		assertThat(out.getHeaders().get("baz"), equalTo("BAZ"));
 	}
 
+	@Test
+	public void testNotPropagateAddWhenNonExist() {
+		AbstractReplyProducingMessageHandler handler = new AbstractReplyProducingMessageHandler() {
+
+			@Override
+			protected Object handleRequestMessage(Message<?> requestMessage) {
+				return new GenericMessage<>("world", Collections.singletonMap("bar", "RAB"));
+			}
+
+		};
+		handler.addNotPropagatedHeaders("boom");
+		assertThat(handler.getNotPropagatedHeaders(), containsInAnyOrder("boom"));
+		handler.setOutputChannel(this.channel);
+		ArgumentCaptor<Message<?>> captor = ArgumentCaptor.forClass(Message.class);
+		willReturn(true).given(this.channel).send(captor.capture());
+		handler.handleMessage(MessageBuilder.withPayload("hello")
+				.setHeader("boom", "FOO")
+				.setHeader("bar", "BAR")
+				.setHeader("baz", "BAZ")
+				.build());
+		Message<?> out = captor.getValue();
+		assertThat(out, notNullValue());
+		assertThat(out.getHeaders().get("boom"), nullValue());
+		assertThat(out.getHeaders().get("bar"), equalTo("RAB"));
+		assertThat(out.getHeaders().get("baz"), equalTo("BAZ"));
```

Comment:
There is a `HeaderMatcher` in the `org.springframework.integration.test.matcher`.
Just FYI, for the future.
We can live with your current test.